storage/s3.go

```diff
@@ -554,11 +581,55 @@ func (s *S3) Put(
 		input.ContentEncoding = aws.String(contentEncoding)
 	}
 
+	// add retry code to the object metadata
+	if s.retryOnNoSuchUploadError > 0 {
+		objectMetaData := make(map[string]*string)
+		num, _ := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+		retryCode = num.String()
+		objectMetaData[retryMetaDataKey] = &retryCode
+		input.Metadata = objectMetaData
+	}
+
 	_, err := s.uploader.UploadWithContext(ctx, input, func(u *s3manager.Uploader) {
 		u.PartSize = partSize
 		u.Concurrency = concurrency
 	})
 
+	i := 0
```

Comment:
We should change the iteration variable name if used after the iteration ends for more readable code. `attempts` maybe?